File: automation-test-playwright/pages/staff/StaffCustomerListPage.ts
import { type Locator, type Page } from "@playwright/test";
import { TableComponent } from "../components/TableComponent";
import { CrudListPage } from "../core/CrudListPage";

export class StaffCustomerListPage extends CrudListPage {
  protected readonly path = "/staff/customers";
  readonly tableBody: Locator;
  private readonly table: TableComponent;

  /**
   * Initializes this page object.
   */
  constructor(page: Page) {
    super(page);
    this.tableBody = this.page.locator("#customerTableBody");
    this.table = new TableComponent(page, "#customerTableBody");
  }

  async waitForLoaded(): Promise<void> {
    await this.page.waitForURL(/\/staff\/customers/);
    await this.waitForTitleContainsLoose("quan ly khach hang", "customer");
    await this.waitForVisible(this.tableBody);
  }

  async waitForTableData(): Promise<void> {
    await this.table.waitForDataOrEmpty();
  }

  /**
   * Applies the full name filter.
   */
  async filterByFullName(fullName: string): Promise<void> {
    await this.fillFilter("fullName", fullName);
  }

  /**
   * Submits the active filters.
   */
  async submitFilters(): Promise<void> {
    await this.search();
  }

  rowByCustomerName(name: string): Locator {
    return this.table.rowByText(name);
  }

  async openCustomerDetail(name: string): Promise<void> {
    await this.rowByCustomerName(name).locator(".btn-view").click();
  }

  async waitForDetailModalContains(name: string): Promise<void> {
    await this.waitForLocatorText(this.page.locator(".modal.show"), name);
  }
}

File: automation-test-selenium/src/main/java/org/suntower/pages/staff/StaffCustomerListPage.java
package org.suntower.pages.staff;

import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.suntower.pages.components.TableComponent;
import org.suntower.pages.core.CrudListPage;

public class StaffCustomerListPage extends CrudListPage {
  private final By tableBody = css("#customerTableBody");
  private final By detailModal = css(".modal.show");
  private final TableComponent table;

  public StaffCustomerListPage(WebDriver driver) {
    super(driver);
    this.path = "/staff/customers";
    this.table = new TableComponent(driver, "#customerTableBody");
  }

  public void waitForLoaded() {
    waitForUrlContains("/staff/customers");
    waitForVisible(tableBody);
  }

  public void waitForTableData() {
    table.waitForDataOrEmpty();
  }

  public void filterByFullName(String fullName) {
    fill(inputByName("fullName"), fullName);
  }

  public void submitFilters() {
    wait.until(ExpectedConditions.presenceOfElementLocated(css("#filterForm")));
    ((JavascriptExecutor) driver).executeScript("document.querySelector('#filterForm').requestSubmit();");
    waitForTableData();
  }

  public WebElement rowByCustomerName(String name) {
    return table.rowByText(name);
  }

  public void openCustomerDetail(String name) {
    waitForCondition(
        () -> {
          try {
            click(rowByCustomerName(name).findElement(By.cssSelector(".btn-view")));
            return true;
          } catch (RuntimeException ignored) {
            return false;
          }
        },
        "Could not open customer detail for: " + name);
  }

  public void waitForDetailModalContains(String name) {
    waitForVisible(detailModal);
    waitForText(detailModal, name);
  }
}
